pkg/config/defaults.go

```diff
@@ -0,0 +1,257 @@
+	DefaultRktStage1ImagePath    = "/usr/lib/rkt/stage1-images/stage1-coreos.aci"
+
+	CacheDir = ".cache"
+)
+
+func SetDefaults_Viper(v *viper.Viper) {
+	v.SetDefault("dir", DefaultKubeSpawnDir)
+	v.SetDefault("cluster-name", DefaultClusterName)
+	v.SetDefault("container-runtime", DefaultContainerRuntime)
+	v.SetDefault("kubernetes-version", DefaultKubernetesVersion)
+	v.SetDefault("dev", false)
+	v.SetDefault("nodes", 2)
+	v.SetDefault("image", DefaultBaseImage)
+}
+
+func SetDefaults_Kubernetes(cfg *ClusterConfiguration) error {
+	var (
+		kubeletPath, kubeadmPath, kubectlPath string
+		kubeletServicePath, kubeadmDropinPath string
+	)
+
+	cacheDir := path.Join(cfg.KubeSpawnDir, CacheDir)
+
+	if cfg.DevCluster {
+		cfg.KubernetesVersion = "latest"
+		// self-compiled k8s development tree
+		k8sOutputDir, err := utils.GetK8sBuildOutputDir()
+		if err != nil {
+			return errors.Wrap(err, "error getting k8s build output directory")
+		}
+		kubeletPath = path.Join(k8sOutputDir, "kubelet")
+		kubeadmPath = path.Join(k8sOutputDir, "kubeadm")
+		kubectlPath = path.Join(k8sOutputDir, "kubectl")
+
+		k8sBuildAssetDir, err := utils.GetK8sBuildAssetDir()
+		if err != nil {
+			return errors.Wrap(err, "error getting k8s build asset directory")
+		}
+		kubeletServicePath = path.Join(k8sBuildAssetDir, "debs/kubelet.service")
+		kubeadmDropinPath = path.Join(k8sBuildAssetDir, "rpms/10-kubeadm.conf")
+	} else {
+		// from download cache
+		kubeletPath = path.Join(cacheDir, cfg.KubernetesVersion, "kubelet")
+		kubeadmPath = path.Join(cacheDir, cfg.KubernetesVersion, "kubeadm")
+		kubectlPath = path.Join(cacheDir, cfg.KubernetesVersion, "kubectl")
+		kubeletServicePath = path.Join(cacheDir, cfg.KubernetesVersion, "kubelet.service")
+		kubeadmDropinPath = path.Join(cacheDir, cfg.KubernetesVersion, "10-kubeadm.conf")
+	}
+
+	cfg.Copymap = []Pathmap{
+		{Dst: "/usr/bin/kubelet", Src: kubeletPath},
+		{Dst: "/usr/bin/kubeadm", Src: kubeadmPath},
+		{Dst: "/usr/bin/kubectl", Src: kubectlPath},
+		{Dst: "/etc/systemd/system/kubelet.service", Src: kubeletServicePath},
+		{Dst: "/etc/systemd/system/kubelet.service.d/10-kubeadm.conf", Src: kubeadmDropinPath},
+		// NOTE: workaround for making kubelet work with port-forward
+		{Dst: "/usr/bin/socat", Src: path.Join(cacheDir, "socat")},
+	}
+
+	if cfg.DevCluster || utils.CheckVersionConstraint(cfg.KubernetesVersion, ">=1.8.0") {
+		cfg.TokenGroupsOption = "--groups=system:bootstrappers:kubeadm:default-node-token"
+	}
+
+	return nil
+}
+
+func SetDefaults_RuntimeConfiguration(cfg *ClusterConfiguration) error {
+	if cfg.RuntimeConfiguration.Timeout == "" {
+		cfg.RuntimeConfiguration.Timeout = DefaultRuntimeTimeout
+	}
+
+	// NOTE: K8s 1.8 or newer fails to run by default when swap is enabled.
+	// So we should disable the feature with an option "--fail-swap-on=false".
+	if !cfg.DevCluster && utils.CheckVersionConstraint(cfg.KubernetesVersion, "<1.8.0") {
+		cfg.RuntimeConfiguration.FailSwapOn = true
```

Comment:
I have asked this before: can we not disable swap for the machines instead?

Reply:
As for the swap, I think it's hard to expect every user to disable swap. It might be easier to do that if we are only dealing with servers, or if most Linux distros disable swap out of the box. Sadly both are not the case.
So it's kind of nasty, but I would prefer to get it handled in this way on the k8s side.

Reply:
I was thinking of using systemd properties (`systemd-nspawn --property ...`) to disable it for the containers, not for the host system :) On a cursory look, I haven't found a way to do this from the CLI though. So I'm not sure it's possible. IIRC LXD does provide an option for that.